pkg/dispatchcli/cmd/update_api.go

```diff
@@ -0,0 +1,98 @@
+///////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+///////////////////////////////////////////////////////////////////////
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/spf13/cobra"
+	"github.com/vmware/dispatch/pkg/api-manager/gen/client/endpoint"
+	"github.com/vmware/dispatch/pkg/api-manager/gen/models"
+	"github.com/vmware/dispatch/pkg/dispatchcli/i18n"
+)
+
+var (
+	updateAPILong    = "update an api based on json"
+	updateAPIExample = `Example json
+	{
+		"authentication": "public",
+		"enabled": true,
+		"function": "hello-py",
+		"hosts": [
+			"example.com"
+		],
+		"methods": [
+			"GET"
+		],
+		"name": "hello-api",
+		"protocols": [
+			"http",
+			"https"
+		],
+		"tags": null,
+		"uris": [
+			"/hello"
+		]
+	}`
+)
+
+func CallUpdateAPI(input interface{}) error {
+	apiBody := input.(*models.API)
+
+	params := endpoint.NewUpdateAPIParams()
+	params.API = *apiBody.Name
+	params.Body = apiBody
+
+	_, err := apiManagerClient().Endpoint.UpdateAPI(params, GetAuthInfoWriter())
+	if err != nil {
+		return formatAPIError(err, params)
+	}
+
+	return nil
+}
+
+func NewCmdUpdateAPI(out io.Writer, errOut io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "api API_FILE",
+		Short:   i18n.T("Update api"),
+		Long:    updateAPILong,
+		Example: updateAPIExample,
+		Args:    cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := updateAPI(out, errOut, cmd, args)
+			CheckErr(err)
+		},
+	}
+
+	return cmd
+}
+
+func updateAPI(out io.Writer, errOut io.Writer, cmd *cobra.Command, args []string) error {
+	filePath := args[0]
+	api := models.API{}
+
+	if filePath != "" {
```

Comment:
will we only be able to update an api from file? 
I expect we are able to do quick update with some simple flags, just like ``create_api.go``

Reply:
The problem I found is that the cobra library that we are using always sets a default value for all of the flags. In that case I can't tell whether the value was set by the user or the excluded it and the value was populated by the library. I thought about maybe using a string array as the input type for each. That way I can say that the array being empty indicated that the flag was not used. In that case I'm not sure how the current update implementations would handle a mostly empty update object. Would it empty the corresponding fields in the backing store? Should the cli do a get of the actual object before doing an update...

tl;dr: If we want people to make updates to a resource via flags there are several changes that need to be made including potentially a round trip to get the update item before performing an update. If that is the desired functionality, I will add that.

Reply:
you are able to set default values 
``cmd.Flags().StringVarP(&functionName, "func", "f", "", "get all apis for specified function")`` the fourth one is default value "" (empty)

you should only pass flags that is set (which means not the default value) to backend.

to my best knowledge, the backend handler do a get first and then will only updated the field passed from the front-end. this need double check, but definitely you don't need to do a cli get in front-end.

Reply:
@tenczar I think I mislead you yesterday :).  I agree with @seanhuxy, though I think we may actually want to do the get and merge in the client.  This is sort of the difference between PUT/PATCH in APIs. I am however open to discussion here.